### PR TITLE
add Db utils

### PR DESCRIPTION
**Pull request by mattlub**
add db utils.js file, containing custom static methods for Mongoose schema, which error if they don't find anything matching the given id.

add these methods to the models.

relates #71

currently 'failing' some checks because codecov is down I think, can be merged anyway.

### CODE REVIEW

**Review by m4v15 — APPROVED** [review]

**Comment by m4v15** [comment]
Again, methods are registered not used, so coverage has gone done

File: src/db/utils.js [hunk 1]
```diff
@@ -0,0 +1,53 @@
+const createCustomDbError = message => {
+  const err = new Error(message)
+  err.name = 'CustomDbError'
```
Comment (des-des) [inline src/db/utils.js]: Yeah, import this in

File: src/db/utils.js [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+  return err
+}
+
+// static method for Mongoose schema
+const findByIdOrError = function (id) {
+  return new Promise((resolve, reject) => {
```
Comment (des-des) [inline src/db/utils.js]: findById seems to return a promise... The promise wrapper is unessisary

```js
findByIdOrError = id => this.findById(id)
  .then(user => {
    if (!user) {
      return Promise.reject(createCustomDbError('No document matching that id'))
    }
    return Promise.resolve(user)
})
```

https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Global_Objects/Promise

File: src/db/utils.js [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+        }
+        resolve(user)
+      })
+      .catch(reject)
+  })
+}
+
+// static method for Mongoose schema
+const findByIdAndUpdateOrError = function (id, data, options) {
+  return new Promise((resolve, reject) => {
```
Comment (des-des) [inline src/db/utils.js]: same as above

File: src/db/utils.js [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+  return err
+}
+
+// static method for Mongoose schema
+const findByIdOrError = function (id) {
+  return new Promise((resolve, reject) => {
+    this.findById(id)
+      .then(user => {
+        if (!user) {
+          return reject(createCustomDbError('No document matching that id'))
```
Comment (des-des) [inline src/db/utils.js]: string from constants

File: src/db/utils.js [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+  err.name = 'CustomDbError'
+  return err
+}
+
+// static method for Mongoose schema
+const findByIdOrError = function (id) {
+  return new Promise((resolve, reject) => {
+    this.findById(id)
+      .then(user => {
+        if (!user) {
```
Comment (des-des) [inline src/db/utils.js]: Can you check explicitly. Ie `if (user === undefined)` or something

File: src/db/utils.js [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+  })
+}
+
+// static method for Mongoose schema
+const findByIdAndUpdateOrError = function (id, data, options) {
+  return new Promise((resolve, reject) => {
+    this.findByIdAndUpdate(id, data, options)
+      .then(user => {
+        if (!user) {
+          return reject(createCustomDbError('Cannot find document to update'))
```
Comment (des-des) [inline src/db/utils.js]: same as above

File: src/db/utils.js [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+      .catch(reject)
+  })
+}
+
+// static method for Mongoose schema
+const findByIdAndUpdateOrError = function (id, data, options) {
+  return new Promise((resolve, reject) => {
+    this.findByIdAndUpdate(id, data, options)
+      .then(user => {
+        if (!user) {
```
Comment (des-des) [inline src/db/utils.js]: same as above

File: src/db/utils.js [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+        }
+        resolve(user)
+      })
+      .catch(reject)
+  })
+}
+
+// static method for Mongoose schema
+const findByIdAndRemoveOrError = function (id, data, options) {
+  return new Promise((resolve, reject) => {
```
Comment (des-des) [inline src/db/utils.js]: same as above

File: src/db/utils.js [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+    this.findByIdAndRemove(id)
+      .then(user => {
+        if (!user) {
```
Comment (des-des) [inline src/db/utils.js]: same as above

File: src/db/utils.js [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+          return reject(createCustomDbError('Cannot find document to delete'))
```
Comment (des-des) [inline src/db/utils.js]: same as above

File: src/db/utils.js (hunk 5 above)
Comment (des-des) [reply src/db/utils.js]: Also, can you not call this user?

File: src/models/Event.js [hunk 11]
```diff
@@ -20,4 +21,9 @@ const eventSchema = mongoose.Schema(
   }
 )
 
+// methods which throw errors when there's nothing matching the given id
+eventSchema.statics.findByIdOrError = findByIdOrError
```
Comment (des-des) [inline src/models/Event.js]: Okay can you build a helper that does this, ie `addStaticMethods`

File: src/models/Place.js [hunk 12]
```diff
@@ -22,4 +23,9 @@ const placeSchema = mongoose.Schema(
   }
 )
 
+// methods which throw errors when there's nothing matching the given id
+placeSchema.statics.findByIdOrError = findByIdOrError
```
Comment (des-des) [inline src/models/Place.js]: use helper.

That way, when you add a new method, you do not need to do it everywhere..

**Review by des-des — CHANGES_REQUESTED** [review]

**Comment by mattlub** [comment]
cheers @des-des , updated

File: src/db/utils.js [hunk 13]
```diff
@@ -6,44 +6,38 @@ const createCustomDbError = message => {
...
-      .catch(reject)
-  })
+  return this.findById(id)
+    .then(res => {
+      if (!res) {
+        return Promise.reject(createCustomDbError('No document matching that id'))
+      }
+      return Promise.resolve(res)
+    })
+    .catch(err => Promise.reject(err))
```
Comment (des-des) [inline src/db/utils.js]: This does not do anything, you can omit it

File: src/db/utils.js [hunk 14]
```diff
@@ -6,44 +6,38 @@ const createCustomDbError = message => {
...
-      .catch(reject)
-  })
+  return this.findByIdAndUpdate(id, data, options)
+    .then(res => {
+      if (!res) {
+        return Promise.reject(createCustomDbError('Cannot find document to update'))
+      }
+      return Promise.resolve(res)
+    })
+    .catch(err => Promise.reject(err))
```
Comment (des-des) [inline src/db/utils.js]: same

File: src/db/utils.js [hunk 15]
```diff
@@ -6,44 +6,38 @@ const createCustomDbError = message => {
...
-      .catch(reject)
-  })
+  return this.findByIdAndRemove(id)
+    .then(res => {
+      if (!res) {
+        return Promise.reject(createCustomDbError('Cannot find document to delete'))
+      }
+      Promise.resolve(res)
+    })
+    .catch(err => Promise.reject(err))
```
Comment (des-des) [inline src/db/utils.js]: same

File: src/db/utils.js (hunk 13 above)
Comment (des-des) [reply src/db/utils.js]: like: I am gonna catch a rejected promise just so I can reject it again

Just seem mean to the error

**Review by des-des — CHANGES_REQUESTED** [review]

**Comment by des-des** [comment]
@mattlub @m4v15 

expand the points above

image

**Review by m4v15 — APPROVED** [review]

**Comment by des-des** [comment]
@m4v15 @mattlub, ready to go, with one comment.

You need to expand my points above to see it

**Review by des-des — APPROVED** [review]

**Pull request merged**